python/ray/tests/test_basic_5.py

```diff
@@ -119,5 +121,95 @@ def test_internal_kv(ray_start_regular):
         kv._internal_kv_list("@namespace_abc", namespace="n")
 
 
+def get_gcs_memory_used():
+    m = sum([
+        process.memory_info().rss for process in psutil.process_iter()
+        if process.name() in ("gcs_server", "redis-server")
+    ])
+    print(m)
+    return m
+
+
+def function_entry_num(job_id):
+    from ray.ray_constants import KV_NAMESPACE_FUNCTION_TABLE
+    return len(_internal_kv_list(b"IsolatedExports:" + job_id,
+                                 namespace=KV_NAMESPACE_FUNCTION_TABLE)) + \
+        len(_internal_kv_list(b"RemoteFunction:" + job_id,
+                              namespace=KV_NAMESPACE_FUNCTION_TABLE)) + \
+        len(_internal_kv_list(b"ActorClass:" + job_id,
+                              namespace=KV_NAMESPACE_FUNCTION_TABLE))
+
+
+@pytest.mark.skipif(
+    client_test_enabled(),
+    reason="client api doesn't support namespace right now.")
+def test_function_table_gc(call_ray_start):
+    """This test tries to verify that function table is cleaned up
+    after job exits.
+    """
+
+    def f():
+        data = "0" * 1024 * 1024  # 1MB
+
+        @ray.remote
+        def r():
+            nonlocal data
+
+            @ray.remote
+            class Actor:
+                pass
+
+        return r.remote()
+
+    ray.init(address="auto", namespace="b")
+
+    # It should use > 500MB data
+    ray.get([f() for _ in range(500)])
+
+    assert get_gcs_memory_used() > 500 * 1024 * 1024
+    job_id = ray.worker.global_worker.current_job_id.binary()
+
+    ray.shutdown()
+
+    # now check the function table is cleaned up after job finished
+    ray.init(address="auto", namespace="a")
+    wait_for_condition(lambda: function_entry_num(job_id) == 0)
+
+
+@pytest.mark.skipif(
+    client_test_enabled(),
+    reason="client api doesn't support namespace right now.")
+def test_function_table_gc_actor(call_ray_start):
+    """If there is a detached actor, the table won't be cleaned up.
+    """
+    ray.init(address="auto", namespace="a")
+
+    @ray.remote
+    class Actor:
+        def ready(self):
+            return
+
+    # If there is a detached actor, function won't be deleted
+    a = Actor.options(lifetime="detached", name="a").remote()
+    ray.get(a.ready.remote())
+    job_id = ray.worker.global_worker.current_job_id.binary()
+    ray.shutdown()
+
+    ray.init(address="auto", namespace="b")
+    with pytest.raises(Exception):
```

Comment:
`wait_for_condition` will just check and sleep. I think what you mean is:
```
        wait_for_condition(lambda: function_entry_num(job_id) == 1)
```
i didn't check function_entry_num because I feel this is just much simpler. it actually won't be 1 I think. also, there might be other things in the function table.